plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/codeInsight/unwrap/KotlinFunctionParameterUnwrapper.kt

```diff
@@ -3,25 +3,176 @@
+                if (valueArgumentList.parent !is KtCallExpression) return null
+                return valueArgumentList.arguments.lastOrNull()
+            }
+            element.isComma -> {
+                // caret before comma, choose previous argument
+                val previous = element.getPrevSiblingIgnoringWhitespaceAndComments()
+                isCallArgument(previous).ifTrue { previous as KtValueArgument }
+            }
+            element is PsiWhiteSpace -> {
+                // caret before blank, argument could be towards the left or the right
+                val previous = element.getPrevSiblingIgnoringWhitespaceAndComments()
+                if (isCallArgument(previous)) return previous as KtValueArgument
+
+                val next = element.getNextSiblingIgnoringWhitespaceAndComments()
+                if (isCallArgument(next)) return next as KtValueArgument
+
+                null
+            }
+            else -> null
+        }
+    }
+
+    private fun isCallArgument(element: PsiElement?): Boolean {
+        if (element is KtLambdaArgument) return false
+        if (element !is KtValueArgument) return false
+        val argumentList = element.parent as KtValueArgumentList
+        if (argumentList.parent !is KtCallExpression) return false
+
         return true
     }
 
     override fun doUnwrap(element: PsiElement?, context: Context?) {
-        val function = element as? KtCallExpression ?: return
-        val argument = element.valueArguments.firstOrNull()?.getArgumentExpression() ?: return
-        context?.extractFromExpression(argument, function)
-        context?.delete(function)
+        val valueArgument = argumentToUnwrap(element) ?: return
+        val deletionTarget = deletionTarget(valueArgument) ?: return
+        val argument = valueArgument.getArgumentExpression() ?: return
+        context?.extractFromExpression(argument, deletionTarget)
+        context?.delete(deletionTarget)
     }
 
+    private fun deletionTarget(valueArgument: KtValueArgument): KtElement? {
+        var function: KtElement = valueArgument.getStrictParentOfType<KtCallExpression>() ?: return null
+        val parent = function.parent
+        if (parent is KtQualifiedExpression) {
+            function = parent
+        }
+        return function
+    }
+
+    override fun collectAffectedElements(e: PsiElement, toExtract: MutableList<PsiElement>): PsiElement {
+        val valueArgument = argumentToUnwrap(e) ?: return e
+        val deletionTarget = deletionTarget(valueArgument) ?: e
+
+        toExtract.add(valueArgument)
+
+        return deletionTarget
+    }
+
+    override fun getDescription(e: PsiElement): String {
+        // necessary because the base implementation expects to be called for KtElement
+        // but because we support the caret to be on LPAR/COMMA and other tokens this doesn't work
+
+        val target = argumentToUnwrap(e) ?: return "<unknown>" // should not be possible
```

Comment:
```suggestion
        val target = argumentToUnwrap(e) ?: error("Description asked for a non applicable unwrapper")
```

Reply:
If IntelliJ ever reached that error, it wouldn't show any unwrap actions at all, no? I don't think throwing an Exception here is worth it.

Reply:
If you think that something isn't achievable then you should throw an exception. It's the idea of exceptions. Right now, you have unlocalized "<unknown>" string